### ISSUE TITLE
tether::dispatch assertion fails when called on other threads

### ISSUE THREAD

**Issue by amiraeva** [quadrupleslap/tether#32]
I want to use `tether::dispatch` to run code on the main UI thread (#30) as advertised. I seem to hit an assertion failure; I managed to break it down into this minimal example.

```rust
use std::{thread, time::Duration};

fn start() {
	let _ = tether::Window::default();

	thread::spawn(|| {
		println!("about to go boom");
		thread::sleep(Duration::from_secs(3));
		tether::dispatch(|| ()); // boom!
	});
}

fn main() {
	unsafe { tether::start(start) }
}

```

Program output:
```
about to go boom
thread '<unnamed>' panicked at 'assertion failed: initialized.get()', tether/src/lib.rs:231:9
note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
```

The offending code in question:
```rust
/// Run the given function on the main thread.
pub fn dispatch<F: FnOnce() + Send>(f: F) {
    assert_initialized();

    unsafe {
        raw::tether_dispatch(Box::<F>::into_raw(Box::new(f)) as _, Some(execute::<F>));
    }

    unsafe extern "C" fn execute<F: FnOnce() + Send>(data: *mut c_void) {
        Box::<F>::from_raw(data as _)();
    }
}

/// Make sure that we're initialized and on the main thread.
fn assert_initialized() {
    INITIALIZED.with(|initialized| {
        assert!(initialized.get());
    });
}
```

The issue appears to be that when `tether::dispatch` is called, it calls `assert_initialized` on that same thread, when instead the assertion should be checked on the main thread.

Moving the assertion check into `execute` seems to fix the issue. But a more robust fix would include `std::panic::catch_unwind` since the possibility of panicking in foreign code is technically undefined behavior AFAIK.

```rust
/// Run the given function on the main thread.
pub fn dispatch<F: FnOnce() + Send>(f: F) {
   // assert_initialized();

    unsafe {
        raw::tether_dispatch(Box::<F>::into_raw(Box::new(f)) as _, Some(execute::<F>));
    }

    unsafe extern "C" fn execute<F: FnOnce() + Send>(data: *mut c_void) {
        assert_initialized(); // where it belongs
        Box::<F>::from_raw(data as _)();
    }
}
```

**Comment by quadrupleslap** [comment]
You're right, dispatch is meant to be called off the main thread, so it's silly to "make sure that we're initialized and on the main thread." We just want to make sure "that we're initialized".

I'm not sure all the platforms let you dispatch stuff before running `init`, though, so the right thing to do is probably to make a `static INITIALIZED: AtomicBool` in addition to the thread-local `MAIN_THREAD: Cell<bool>`, and check the static one inside dispatch.

**Comment by amiraeva** [comment]
Thanks - I created a PR which (I think) fixes this using your suggestion!

**Comment by quadrupleslap** [comment]
Closed by #4623e99.
